From 6b09c57b477e7cd5c18133b38c26788dcb2aec57 Mon Sep 17 00:00:00 2001
From: Valentin Caron <valentin.caron@foss.st.com>
Date: Mon, 25 Sep 2023 11:17:06 +0200
Subject: [PATCH 0554/1141] tty: Revert flow-control char function on closing
 path

This reverts commit 65534736d9a5cab5340ae8819e1394b6325e8390.

This series seems to introduce some regressions on tty buffer.
Revert it temporary.

Signed-off-by: Valentin Caron <valentin.caron@foss.st.com>
Change-Id: I7922c9112b65f3ef140efbc2475ba3c01e27408e
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/329488
Reviewed-by: Amelie DELAUNAY <amelie.delaunay@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Amelie DELAUNAY <amelie.delaunay@foss.st.com>
---
 drivers/tty/n_tty.c | 13 +++++++++----
 1 file changed, 9 insertions(+), 4 deletions(-)

diff --git a/drivers/tty/n_tty.c b/drivers/tty/n_tty.c
index 4dff2f34e2d0..ad83ce5f03cb 100644
--- a/drivers/tty/n_tty.c
+++ b/drivers/tty/n_tty.c
@@ -1434,10 +1434,15 @@ static void n_tty_receive_char_closing(struct tty_struct *tty, unsigned char c,
 		c = tolower(c);
 
 	if (I_IXON(tty)) {
-		if (!n_tty_receive_char_flow_ctrl(tty, c, lookahead_done) &&
-		    tty->flow.stopped && !tty->flow.tco_stopped && I_IXANY(tty) &&
-		    c != INTR_CHAR(tty) && c != QUIT_CHAR(tty) &&
-		    c != SUSP_CHAR(tty)) {
+		if (c == STOP_CHAR(tty)) {
+			if (!lookahead_done)
+				stop_tty(tty);
+		} else if (c == START_CHAR(tty) && lookahead_done) {
+			return;
+		} else if (c == START_CHAR(tty) ||
+			 (tty->flow.stopped && !tty->flow.tco_stopped && I_IXANY(tty) &&
+			  c != INTR_CHAR(tty) && c != QUIT_CHAR(tty) &&
+			  c != SUSP_CHAR(tty))) {
 			start_tty(tty);
 			process_echoes(tty);
 		}
-- 
2.39.2

